src/server.py

```diff
@@ -0,0 +1,188 @@
+AGENT_JAR_PATH = Path(JENKINS_WORKDIR / "agent.jar")
+AGENT_READY_PATH = Path(JENKINS_WORKDIR / "agents/.ready")
+ENTRYSCRIPT_PATH = Path(JENKINS_WORKDIR / "entrypoint.sh")
+
+USER = "jenkins"
+GROUP = "jenkins"
+
+
+class Credentials(BaseModel):
+    """The credentials used to register to the Jenkins server.
+
+    Attrs:
+        address: The Jenkins server address to register to.
+        secret: The secret used to register agent.
+    """
+
+    address: str
+    secret: str
+
+    @classmethod
+    def from_jenkins_slave_interface_dict(
+        cls, server_unit_databag: ops.RelationDataContent
+    ) -> typing.Optional["Credentials"]:
+        """Import server metadata from databag in slave relation.
+
+        Args:
+            server_unit_databag: The relation databag content from slave relation.
+
+        Returns:
+            Metadata if complete values(url, secret) are set. None otherwise.
+        """
+        address = server_unit_databag.get("url")
+        secret = server_unit_databag.get("secret")
+        if not address or not secret:
+            return None
+        return Credentials(address=address, secret=secret)
+
+    @classmethod
+    def from_jenkins_agent_v0_interface_dict(
+        cls, server_unit_databag: ops.RelationDataContent, unit_name: str
+    ) -> typing.Optional["Credentials"]:
+        """Import server metadata from databag in agent relation.
+
+        Args:
+            server_unit_databag: The relation databag content from agent relation.
+            unit_name: The agent unit name.
+
+        Returns:
+            Metadata if complete values(url, secret) are set. None otherwise.
+        """
+        address = server_unit_databag.get("url")
+        secret = server_unit_databag.get(f"{unit_name}_secret")
+        if not address or not secret:
+            return None
+        return Credentials(address=address, secret=secret)
+
+
+def get_credentials(
+    relation_name: str, unit_name: str, databag: typing.Optional[ops.RelationDataContent]
+) -> typing.Optional["Credentials"]:
+    """Get credentials from databag.
+
+    Args:
+        relation_name: The relation name of the databag. Either "slave" or "agent".
+        unit_name: The agent unit name.
+        databag: The relation databag for given relation.
+
+    Returns:
+        Credentials if databag contains valid metadata. None if partial or no metadata found.
+    """
+    if databag is None:
+        return None
+    if relation_name == state.SLAVE_RELATION:
+        return Credentials.from_jenkins_slave_interface_dict(databag)
+    return Credentials.from_jenkins_agent_v0_interface_dict(databag, unit_name=unit_name)
+
+
+class ServerBaseError(Exception):
+    """Represents errors with interacting with Jenkins server."""
+
+
+class AgentJarDownloadError(ServerBaseError):
+    """Represents an error downloading agent JAR executable."""
+
+
+def download_jenkins_agent(server_url: str, connectable_container: ops.Container) -> None:
```

Comment:
Does this need to know it is connectable? Shouldn't it just be container and assume that it is ready to be interacted with?

Reply:
changed to be just container, removed all other instances of `connectable_container`. Thank you!